scripts/asr_language_modeling/neural_rescorer/eval_neural_rescorer.py

```diff
@@ -0,0 +1,226 @@
+    else:
+
+        @contextlib.contextmanager
+        def autocast():
+            yield
+
+    logging.info(f"Rescoring with beam_size: {args.beam_size}")
+    logging.info("Calculating the scores...")
+    with autocast():
+        with torch.no_grad():
+            am_scores, lm_scores, dists, ref_lens, lens_in_chars = [], [], [], [], []
+            for batch in tqdm.tqdm(data_loader):
+                input_ids, input_mask, acoustic_score, dist, ref_len, len_in_chars, idx = batch
+
+                max_len_in_batch = input_mask.sum(dim=0).argmin().item()
+                input_ids, input_mask = input_ids[:, :max_len_in_batch], input_mask[:, :max_len_in_batch]
+                if torch.cuda.is_available():
+                    input_ids, input_mask = input_ids.to(device), input_mask.to(device)
+                    dist, acoustic_score, len_in_chars = (
+                        dist.to(device),
+                        acoustic_score.to(device),
+                        len_in_chars.to(device),
+                    )
+
+                log_probs = model.forward(input_ids[:, :-1], input_mask[:, :-1])
+                target_log_probs = log_probs.gather(2, input_ids[:, 1:].unsqueeze(2)).squeeze(2)
+                neural_lm_score = torch.sum(target_log_probs * input_mask[:, 1:], dim=-1)
+
+                am_scores.append(acoustic_score)
+                lm_scores.append(neural_lm_score)
+                dists.append(dist)
+                ref_lens.append(ref_len)
+                lens_in_chars.append(len_in_chars)
+
+    am_scores = torch.cat(am_scores).view(-1, args.beam_size)
+    lm_scores = torch.cat(lm_scores).view(-1, args.beam_size)
+    dists = torch.cat(dists).view(-1, args.beam_size)
+    ref_lens = torch.cat(ref_lens).view(-1, args.beam_size)
+    lens_in_chars = torch.cat(lens_in_chars).view(-1, args.beam_size).to(am_scores.dtype)
+
+    total_len = ref_lens[:, 0].sum()
+    model_wer = dists[:, 0].sum() / total_len
+    ideal_wer = dists.min(dim=1)[0].sum() / total_len
+
+    if args.alpha is None:
+        logging.info("Linear search for alpha...")
+        coef1, _ = line_search_wer(dists=dists, scores1=am_scores, scores2=lm_scores, total_len=total_len)
+        coef1 = np.round(coef1, 3)
+        logging.info(f"alpha={coef1} achieved the best WER.")
+        logging.info(f"------------------------------------------------")
+    else:
+        coef1 = args.alpha
+
+    scores = am_scores + coef1 * lm_scores
+
+    if args.beta is None:
+        logging.info("Linear search for beta...")
+        coef2, _ = line_search_wer(dists, scores, lens_in_chars, total_len)
+        coef2 = np.round(coef2, 3)
+        logging.info(f"beta={coef2} achieved the best WER.")
+        logging.info(f"------------------------------------------------")
+    else:
+        coef2 = args.beta
+
+    ab_scores = am_scores + coef1 * lm_scores + coef2 * lens_in_chars
+    ab_wer = compute_wer(dists, ab_scores, total_len)
+
```

Comment:
Would it be possible to cache the results of neural rescoring so that it doesn't need to be recomputed? Or is it fast enough that it's not needed ?

Reply:
For instance, it would take around 2 mins for test-other LS, so no need for caching. Also user usually need to run it once or twice.

Reply:
That's plenty fast, no point caching unlike the beam search which takes modest time